substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -1425,4 +1499,75 @@ impl<T: Config> ChangeNode for Module<T> {
             Self::remove_contract(rent_contract.contract_id);
         }
     }
+
+    pub fn _create_solution_provider(
+        description: Vec<u8>,
+        link: Vec<u8>,
+        providers: Vec<types::Provider<T::AccountId>>,
+    ) -> DispatchResult {
+        let total_take: u8 = providers.iter().map(|provider| provider.take).sum();
+        ensure!(total_take <= 50, Error::<T>::InvalidProviderConfiguration);
+
+        let mut id = SolutionProviderID::get();
+        id = id + 1;
+
+        let solution_provider = types::SolutionProvider {
+            solution_provider_id: id,
+            providers,
+            description,
+            link,
+            approved: false,
+        };
+
+        SolutionProviderID::put(id);
+        SolutionProviders::<T>::insert(id, &solution_provider);
+
+        Self::deposit_event(RawEvent::SolutionProviderCreated(solution_provider));
+
+        Ok(())
+    }
+
+    pub fn _approve_solution_provider(solution_provider_id: u64, approve: bool) -> DispatchResult {
+        ensure!(
+            SolutionProviders::<T>::contains_key(solution_provider_id),
+            Error::<T>::NoSuchSolutionProvider
+        );
+
+        let mut solution_provider = SolutionProviders::<T>::get(solution_provider_id);
+        solution_provider.approved = approve;
+        SolutionProviders::<T>::insert(solution_provider_id, &solution_provider);
+
+        Self::deposit_event(RawEvent::SolutionProviderApproved(
+            solution_provider_id,
+            approve,
+        ));
+
+        Ok(())
+    }
+
+    pub fn validate_solution_provider(solution_provider_id: Option<u64>) -> DispatchResult {
+        match solution_provider_id {
+            Some(provider_id) => {
+                ensure!(
+                    SolutionProviders::<T>::contains_key(provider_id),
+                    Error::<T>::NoSuchSolutionProvider
+                );
+                let solution_provider = SolutionProviders::<T>::get(provider_id);
+                ensure!(
+                    solution_provider.approved,
+                    Error::<T>::SolutionProviderNotApproved
+                );
+                return Ok(());
+            }
+            None => return Ok(()),
+        };
+    }
+}
```

Comment:
Since the `None` case doesn't need handling, this can be rewritten using `if let Some(provider_id) = solution_provider_id { ...`. Also both cases have the same return, so it can be omitted in the body and just written as an implicit return statement on the last line (`Ok(())`)